v2/api/cache/customizations/redis_extension_types.go

```diff
@@ -0,0 +1,47 @@
+/*
+Copyright (c) Microsoft Corporation.
+Licensed under the MIT license.
+*/
+
+package customizations
+
+import (
+	"strings"
+
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/go-logr/logr"
+
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime/core"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime/extensions"
+)
+
+var _ extensions.ErrorClassifier = &RedisExtension{}
+
+// ClassifyError evaluates the provided error, returning including whether it is fatal or can be retried.
+// A conflict error (409) is normally fatal, but Redis resources may return 409 whilst a dependency is being created,
+// so we override for that case.
+// cloudError is the error returned from ARM.
+// apiVersion is the ARM API version used for the request.
+// log is a logger than can be used for telemetry.
+// next is the next implementation to call.
+func (e *RedisExtension) ClassifyError(
+	cloudError *core.CloudError,
+	apiVersion string,
+	log logr.Logger,
+	next extensions.ErrorClassifierFunc) (core.CloudErrorDetails, error) {
```

Comment:
@Porges suggested the model that I adopted here. 

Passing the default logic as `next` allows extensions to work in four different ways:

* Add custom behavior only _before_ the default
* Add custom behaviour only _after_ the default 
* Wrap the default behaviour in custom logic
* Completely replace the functionality (ignoring `next`)

The first two are probably going to cover most cases (though not all) and failing to provide easy access to the default behaviour risks extensions reinventing the wheel. If they do so in a way accidentally inconsistent with the default behaviour, we might get some surpsing edge cases that are hard to diagnose. We end up with the same situation if they _copy_ the default behaviour (violating DRY), and then we fix a bug in the original.

> since the purpose of the extension is to override the default

That's the fourth case listed above; I think it's going to be far more common for extensions to _augment_ the default.

Reply:
I was expecting that the extension for error handling would always make a decision based on the response from Azure. There would be no need for "before or after", because the extension isn't making it's choice based on what the default classifier does, it's overriding _part_ of the classifier (your augment case). Then if the extension didn't make any classification, the default would apply.

Something like this:

```
type ErrorClassifier interface {
     func ClassifyError(
         cloudError *genericarmclient.CloudError,
        apiVersion string,
        log logr.Logger) (*genericarmclient.CloudErrorDetails, error)
}

func (e *RedisExtension) ClassifyError(
    cloudError *genericarmclient.CloudError,
    apiVersion string,
    log logr.Logger) (genericarmclient.CloudErrorDetails, error) {

    if cloudError.Inner != nil && cloudError.Inner.Code == "Conflict" && inner.Message != nil && strings.Contains(strings.ToLower(*inner.Message), "try again later") {
        return core.MakeCloudErrorDetails(cloudError, classification) // This is a new method that extracts the code/message to their respective fields
    }

    return nil
}

func CreateErrorClassifier(
    host genruntime.ResourceExtension,
    classifier ErrorClassifierFunc,
    apiVersion string,
    log logr.Logger) ErrorClassifierFunc {
    impl, ok := host.(ErrorClassifier)
    if !ok {
        return classifier
    }

    return func(cloudError *genericarmclient.CloudError) (genericarmclient.CloudErrorDetails, error) {
        log.V(Info).Info(
            "Classifying cloud error",
            "Message", cloudError.InnerError.Message,
            "Code", cloudError.InnerError.Code,
            "Target", cloudError.InnerError.Target)

        result, err := impl.ClassifyError(cloudError, apiVersion, log, classifier)
        if err != nil {
            return genericarmclient.CloudErroDetails{}, err
        }
        if result == nil {
            result, err = classifier(cloudError)
        }

        log.V(Info).Info(
            "Cloud error classified",
            "Classification", result.Classification,
            "Code", result.Code,
            "Message", result.Message)
        return result, err
    }
}
```

I'm ok with it being more powerful, I just think it could lead people to write code like you did where you're checking "what did the default classifier do": [if details.Classification == genericarmclient.ErrorFatal](https://github.com/Azure/azure-service-operator/pull/2099/files#diff-420d9866ee41deef2e288d2da9accce7f98f9fed82823b633f68695db21c3829R38), which feels weird to me. Why does the extension care what the default classifier classified this error as? It _knows_ that this error is nonfatal in this context, why does it matter how the `next` function classified it?

With all that said, this is pretty minor! Maybe I'm just not seeing the cases where the actual extensions will care about what the default classifier thought and only want to make modifications based on that.

Reply:
I took out the check to see what the default classifier did as it was redundant.

> I'm just not seeing the cases where the actual extensions will care about what the default classifier thought and only want to make modifications based on that.

There are two cases. The first is when the error classification extension needs to do something we haven't anticipated. The other is that I'd like most of the extension points to have parallel structure, and we can foresee extensions where they might want to do that (or might want to _sometimes_ skip the default behaviour completely).